Plugins: Allow apps to expose components. Update the extensions API.

leventebalogh opened the pull request:
**Related issue / epic:** https://github.com/grafana/grafana/issues/77856

## What changed?

### Exposable components
This PR introduces something called "exposable components", that can be used to share React components between plugins: 
```ts
// Expose a component (in the `module.tsx` file inside your plugin.)
// You have to prefix the `id` with your plugin id. 
// (Other plugins can use `usePluginComponent(id)` to get hold of the exposed component). 
// -----------------------
.exposeComponent({
    id: 'myorg-foo-app/visualisation-component'
    title: 'Visualisation component',
    description: '...',
    component: () => (
      <div>Hello world!</div>
    )
})


// Accessing a component 
// (either in your plugin or in core Grafana) 
// -------------------
import { usePluginComponent } from "@grafana/runtime";

export const Page = () => {
  const { component: PredictComponent, isLoading } = usePluginComponent("grafana-ml-app/predict");

  return <PredictComponent />;
}
```

### API changes
We did the following changes to the app plugin API:
```ts
// Add a component to an extension point 
// (use this instead of `.configureExtensionComponent()`)
.addComponent({
  targets: ["...", "..."],
  title: "...",
  component: MyComponent
})

// Add a link to an extension point 
// (use this instead of `.configureExtensionLink()`)
.addLink({
  targets: ["...", "..."],
  title: "...",
  onClick: () => {...},
})

// DEPRECATED
// (use `addComponent()` instead)
.configureExtensionComponent({
  extensionPointId: "...",
  title: "...",
  component: MyComponent
})

// DEPRECATED
// (use `addLink()` instead)
.configureExtensionLink({
  extensionPointId: "...",
  title: "...",
  path: "..."
})
```

## Todo:
- [x] Add tests
- [x] Clean up names
- [x] Add an reactive hook

yduartep commented:
@leventebalogh what is the difference between this `configureExtensionFunction` and the existing `configureExtensionComponent` that is already part of the `plugin extension` feature? are they similar ways to extend the UI?

leventebalogh commented:
Hey @yduartep 👋  Very good question! This PR is still in draft, and the "function extensions" are only an experiment at this point, they might won't land in the end. The reasoning and the difference: although react components can be functions as well (function components), for those extensions we have stricter type restrictions than for ordinary functions. Another difference is that function extensions can be async functions, which can be useful in use cases like for example expecting search results from plugins for a certain query string. 

I hope it answered your question - please let me know if it didn't. 
We will make sure that they have proper documentation in case they land in our public API.

yduartep commented:
> Hey @yduartep 👋 Very good question! This PR is still in draft, and the "function extensions" are only an experiment at this point, they might won't land in the end. The reasoning and the difference: although react components can be functions as well (function components), for those extensions we have stricter type restrictions than for ordinary functions. Another difference is that function extensions can be async functions, which can be useful in use cases like for example expecting search results from plugins for a certain query string.
> 
> I hope it answered your question - please let me know if it didn't. We will make sure that they have proper documentation in case they land in our public API.

Thank you. Just try to explain very good this on the final documentation and if is possible include some real use cases to help to understand the feature.

marefr commented:
Nit. We've been talking about plugin capabilities in a greater scope of what plugins can do/contributes to so maybe a bit confusing naming this "capability"? What about command, ref https://code.visualstudio.com/api/extension-capabilities/common-capabilities#command

yduartep commented:
Hi @leventebalogh when do you think this `configureCapability` will be in prod? I would like to start working with the new functionality that is amazing!!

leventebalogh commented:
Hi @yduartep 👋 

We are working very hard to get this feature/branch into a final version and merge it by this Friday. We will keep you in the loop :)

leventebalogh committed:
feat: introduce exposable components and update the public APIs

Co-authored-by: Marcus Andersson <marcus.andersson@grafana.com>

leventebalogh committed:
tests: fix the tests for `usePluginComponent()`

I broke them when I wrapped the component with the PluginContextProvider which fetches the plugin metadata.

leventebalogh committed:
fix: typo

leventebalogh commented:
/deploy-to-hg

jackw reviewed (approved):
LGTM! 🚀
